data/src/main/java/com/linkedin/data/schema/compatibility/AnnotationCompatibilityChecker.java

```diff
@@ -0,0 +1,176 @@
+ *  they may also provide a annotation compatibility method to define the way to check annotation's compatibility.
+ *  In this checker, it will call the {@link SchemaAnnotationHandler#checkCompatibility} method.
+ *
+ * @author Yingjie Bi
+ */
+public class AnnotationCompatibilityChecker
+{
+  private static final Logger _log = LoggerFactory.getLogger(AnnotationCompatibilityChecker.class);
+
+  /**
+   * Check the pegasus schema annotation compatibility
+   * Process prevSchema and currSchema in the SchemaAnnotationProcessor to get the resolved result with resolvedProperties.
+   * then using the resolvedProperties to do the annotation compatibility check.
+   * @param prevSchema previous data schema
+   * @param currSchema current data schema
+   * @param handlers SchemaAnnotationHandler list
+   * @return List<AnnotationCompatibilityResult>
+   */
+  public static List<AnnotationCompatibilityResult> checkPegasusSchemaAnnotation(DataSchema prevSchema, DataSchema currSchema,
+      List<SchemaAnnotationHandler> handlers)
+  {
+    SchemaAnnotationProcessor.SchemaAnnotationProcessResult prevSchemaResult = processSchemaAnnotation(prevSchema, handlers);
+    SchemaAnnotationProcessor.SchemaAnnotationProcessResult currSchemaResult = processSchemaAnnotation(currSchema, handlers);
+    Map<PathSpec, Pair<CompatibilityCheckContext, Map<String, Object>>> prevResolvedPropertiesMap
+        = getNodeToResolvedProperties(prevSchemaResult);
+    Map<PathSpec, Pair<CompatibilityCheckContext, Map<String, Object>>> currResolvedPropertiesMap
+        = getNodeToResolvedProperties(currSchemaResult);
+
+    return getCompatibilityResult(prevResolvedPropertiesMap, currResolvedPropertiesMap, handlers);
+  }
+
+  /**
+   * Iterate the nodeToResolverPropertiesMap, if a node's resolvedProperty contains the same annotationNamespace as SchemaAnnotationHandler,
+   * calling annotationCompatibilityCheck api which is provided in the SchemaAnnotationHandler to do the annotation compatibility check.
+   */
+  private static List<AnnotationCompatibilityResult> getCompatibilityResult(Map<PathSpec, Pair<CompatibilityCheckContext,
+      Map<String, Object>>> prevResolvedPropertiesMap, Map<PathSpec, Pair<CompatibilityCheckContext,
+      Map<String, Object>>> currResolvedPropertiesMap, List<SchemaAnnotationHandler> handlers)
+  {
+    List<AnnotationCompatibilityResult> results = new ArrayList<>();
+    for (SchemaAnnotationHandler handler : handlers)
+    {
+      if (!handler.implementsCheckCompatibility())
+      {
+        _log.info("There is no implementation of checkCompatibility method in SchemaAnnotationHandler: "
+            + handler.getClass().getName());
+        continue;
+      }
+      String annotationNamespace = handler.getAnnotationNamespace();
+      Map<PathSpec, Pair<CompatibilityCheckContext,
+          Map<String, Object>>> copyOfCurrResolvedPropertiesMap = new HashMap<>(currResolvedPropertiesMap);
```

Comment:
This might be a problem for large schemas. Did you consider reversing the loops?

```
remove handlers that don't implement compatibility

foreach entry in prevResolvedProperties:
  foreach hander:
     check(entry)
     remove path from currProperties
foreach entry in currProperties
   foreach handler:
      check(entry)
```

Reply:
Refactored this part based on the recommendation.